### ISSUE TITLE
Random hangs

### ISSUE THREAD

**Issue by coldlamper** [ncb000gt/node-cron#141]
After updating to 1.0.6, the cron will randomly hang and not fire. I have it set to run every second and at random times, hours in, it will just hang and the cron will not fire anymore.  It's hard to reproduce except waiting for a long period of time. Maybe you could possibly see a scenario in which that could happen.

```
var cronJob = require('cron').CronJob;
moment = require('moment-timezone');
var count = 0;

var job = new cronJob('*/1 * * * * *', function () {
	count++;
	console.log('Count: ' + count);
	
        // read and write from redis 
});

job.start();
```

It will end up stuck on:
...
Count: 14565
Count: 14566
Count: 14567

or similiar

**Comment by ncb000gt** [comment]
Interesting, I'll take a look when I get a chance. Thanks.

**Comment by goldfire** [comment]
I can confirm that we are seeing the same issue. We have a cron that runs every 10 seconds that we started logging and since upgrading to 1.0.6 it will just randomly stop firing after a few hours. There are no error messages or anything else I can see to explain why it stops.

**Comment by alihalabyah** [comment]
Looks like that's true

**Comment by ncb000gt** [comment]
Ok, thanks for the notes.

**Comment by ncb000gt** [comment]
I'm still having issues reproducing this issue...what version of node are you on?

    var CronJob = require('./lib/cron').CronJob;
    var sinon = require('sinon');
    var clock = sinon.useFakeTimers();
    var previous = 0;
    var counter = 0;
    var cronSyncSchedule = '* * * * * *';
    
    var scheduleJob = new CronJob(cronSyncSchedule, function() {
        counter++;
        checkCounter();
    }, function() {
    }, true, null);
    clock.tick(24*60*60*1000);
    
    function checkCounter() {
        //console.log(counter + ' - ' + previous);
        if (counter != (previous + 1))
                console.log('SKIPPED: ' + (counter - 1) + ", COUNTER: " + counter);
        previous = counter;
    }
    
    clock.restore();

Also, would you agree this is an acceptable test to try to reproduce what you're seeing (simulated onTicks that is). The reason for simulated is that if it's a problem with the module then the onTicks should not work as expected. If it's a problem with timers or something similar then that's probably a node/iojs issue...

**Comment by goldfire** [comment]
We are using Node 0.10.36 and the only difference was going from 1.0.5 to 1.0.6 and then back to fix it. The cron job that was having the issues for us was this one:

```javascript
new CronJob('*/10 * * * * *', function() {
  console.log('onTick');
  // ....
}, null, true);
```

This would run for several hours and then would just stop firing (I don't know if it was the same length of time each time).

**Comment by coldlamper** [comment]
We are using Node 0.10.33. Reverting back to 1.0.5 fixes the issue.
I tried your test and could not reproduce the error like that.

**Comment by ncb000gt** [comment]
Ok. I'll see if I can try to reproduce it here.

**Comment by coldlamper** [comment]
I ran your test without sinon simulated clock ticks modified like:
```javascript

var CronJob = require('cron').CronJob;

//var sinon = require('sinon');

//var clock = sinon.useFakeTimers();
var previous = 0;
var counter = 0;
var cronSyncSchedule = '*/1 * * * * *';

var scheduleJob = new CronJob(cronSyncSchedule, function() {
	counter++;
	checkCounter();
}, function() {
}, true, null);
//clock.tick(24*60*60*1000);

function checkCounter() {

	console.log(counter + ' - ' + previous);
	if (counter != (previous + 1))
		console.log('SKIPPED: ' + (counter - 1) + ", COUNTER: " + counter);
	previous = counter;
}

//clock.restore();
```

It ended up stuck on
...
18999 - 18998
18900 - 18999
18901 - 18900

It didn't exit, just stopped.

**Comment by keqinzhao** [comment]
I have the same problem. it will stop after run for about 16000 times.

**Comment by ncb000gt** [comment]
Does the incomplete function fire for any of you? Also, are any of you
running the latest stable node?

**Comment by ncb000gt** [comment]
Oncomplete not incomplete...

**Comment by godspeedelbow** [comment]
*+1* experienced this last night, we are on `1.0.6` and `0.10.33`

**Comment by colorcyan** [comment]
+1 , interestingly from our log, we had a rule for 00 * * * * *, it kept firing at exact **:**:00 for 3 days until the last event when it stoppped with the last event logged at timestamp **:**:01

**Comment by jdarling** [comment]
I have been having a similar issue in my own code and was about to switch to Cron when I saw this defect.  From what I can tell this is something to do with the latest versions of Node and the way setTimeout/nextTick work.  Deep inside error reports you will get something about recursion.  Apparently setTimeout isn't a tailed method anymore.

To work around try wrapping all setTimeout's in setImmediate:

```
return setImmediate(function(){
  setTimeout(whatever, whenever);
});
```

At least this seems to work.

**Comment by h0x91b** [comment]
Confirm, after several hours some of my crons stops working

**Comment by h0x91b** [comment]
Cron stops working in function start() if fallen in 

if(timeout >= 0) { ... } else {
//there....
this.stop();
}

because sometime cronTime.getTimeout() return -1 for some reason
var timeout = this.cronTime.getTimeout();
is -1

**Comment by ncb000gt** [comment]
I just pushed a new version of the module with GH-141, this should take care of the hang issues you're seeing. Please post back if not and I can reopen.

**Comment by coldlamper** [comment]
Unfortunately this issue is occurring again using node 0.12.5 and node-cron 1.0.9.  Running the crons from 1  to 2 sec intervals like */2 * * * * *
The cron just stops firing and the program just hangs until i ctrl-c out.
Seems to occur more when system is under a moderate to heavy load.  I would estimate it's been happening 1-3 times a week for at least the last month.

**Comment by medisoft** [comment]
Is this fixed on 1.3? On 1.1 I'm still having the problem.
